File: sylee723/4주차/PGS_주식가격.java
import java.util.*;

class Solution {
    public int[] solution(int[] prices) {
        Stack<int[]> stack = new Stack<>();
        int[] answer = new int[prices.length];
        
        for (int i = 0; i < prices.length; i++) {
            while (!stack.isEmpty() && stack.peek()[1] > prices[i]) {
                int[] pop = stack.pop();
                answer[pop[0]] = i - pop[0];
            }
            
            stack.push(new int[] {i, prices[i]});
        }
        
        while (!stack.isEmpty()) {
            int[] pop = stack.pop();
            answer[pop[0]] = prices.length - 1 - pop[0];
        }
        
        return answer;
    }
}

File: daehwa_park/14주차/PGS_주식가격.py
from collections import deque

def solution(prices):
    answer = [0] * len(prices)
    s = deque()
    
    time = 0
    for i, p in enumerate(prices):
        if not s:
            s.append([p, i])
        else:
            while s and s[-1][0] > p:
                cp, ci = s.pop()
                answer[ci] = time - ci
            s.append([p, i])
        time += 1
    
    while s:
        cp, ci = s.pop()
        answer[ci] = time - ci - 1
    
    return answer